### ISSUE TITLE
OS X 10.15.7 examples fail

### ISSUE THREAD

**Issue by keleher** [hanwen/go-fuse#377]
I have OSXFuse 3.11.2 on a recent iMac.

```
titan:~/go/src/github.com/hanwen/go-fuse/example/zipfs> go run main.go /tmp/z foo.zip 
# github.com/hanwen/go-fuse/v2/zipfs
../../zipfs/zipfs.go:88:5: out.Blksize undefined (type *fuse.AttrOut has no field or method Blksize)
titan:~/go/src/github.com/hanwen/go-fuse/example/zipfs> 
```

and 

```
titan:~/go/src/github.com/hanwen/go-fuse/example/hello> go run main.go tmp
2020/10/16 13:08:11 Mount fail: result too large
exit status 1
titan:~/go/src/github.com/hanwen/go-fuse/example/hello> 
```

**Comment by jim3ma** [comment]
+1

**Comment by jim3ma** [comment]
After update fuse/poll_darwin.go, it works
```
diff --git a/fuse/poll_darwin.go b/fuse/poll_darwin.go
index 84cc980..8a0868d 100644
--- a/fuse/poll_darwin.go
+++ b/fuse/poll_darwin.go
@@ -32,7 +32,7 @@ func pollHack(mountPoint string) error {
                POLLHUP   = 0x10
        )
 
-       fd, err := syscall.Open(filepath.Join(mountPoint, pollHackName), syscall.O_CREAT|syscall.O_TRUNC|syscall.O_RDWR, 0644)
+       fd, err := syscall.Open(filepath.Join("/tmp", pollHackName), syscall.O_CREAT|syscall.O_TRUNC|syscall.O_RDWR, 0644)
        if err != nil {
                return err
        }
```

**Comment by Thireus** [comment]
What about the following:

```
fd, err := syscall.Open(filepath.Join(filepath.Join(mountPoint, "/.."), pollHackName+"."+filepath.Base(mountPoint)), syscall.O_CREAT|syscall.O_TRUNC|syscall.O_RDWR, 0644)
```

That creates the .go-fuse-epoll-hack in the directory where the mountPoint is and would it it the following name: .go-fuse-epoll-hack.{mountPoint}.

**Comment by hanwen** [comment]
the pollHack file must be located inside the mount point, or else it won't trigger the POLL opcode, which is the objective of the entire exercise. 

The "result too large" suggests something else is wrong. What is the debug output look like for the failure case?

**Comment by Thireus** [comment]
I'm sorry, I don't know what pollHack is for. Where can I read more about POLL opcode?

`fs.Mount failed: result too large` <-- That's really the only output. @keleher describes the issue accurately with the hello example.

Code 34 ERANGE. According to: https://bbengfort.github.io/snippets/2017/01/23/syscall-errno.html

**Comment by hanwen** [comment]
CAn you try the patch at https://review.gerrithub.io/c/hanwen/go-fuse/+/506915 ?

for background on the poll hack, see 4f10e248ebabd3cdf9c0aa3ae58fd15235f82a79

**Comment by Thireus** [comment]
Thank you. I've tried, same issue.

I've added the following print to your patch to make sure it's executed:
```
	case _OP_ACCESS, _OP_FLUSH, _OP_RELEASE:
		// Avoid upsetting the OSX mount process.
		fmt.Println("THIREUS: IS THIS EVEN EXECUTED?")
		req.status = OK
```

This is what I've got:
```
20:57:59.982084 rx 2: INIT n0 {7.19 Ra 0x200000 VOL_RENAME,CASE_INSENSITIVE,XTIMES,ATOMIC_O_TRUNC,0x18000000} 
20:57:59.982175 tx 2:     OK, {7.8 Ra 0x200000  0/0 Wr 0x20000 Tg 0x0}
20:57:59.982260 rx 2: STATFS n1 
20:57:59.982268 rx 3: STATFS n1 
20:57:59.982333 tx 2:     OK, {blocks (54541,54286)/1571844 files 16092297653/16095688400 bs1048576 nl0 frs1048576}
20:57:59.982346 tx 3:     OK, {blocks (54541,54286)/1571844 files 16092297653/16095688400 bs1048576 nl0 frs1048576}
20:57:59.982597 rx 2: STATFS n1 
20:57:59.982615 rx 3: STATFS n1 
20:57:59.982635 tx 2:     OK, {blocks (54541,54286)/1571844 files 16092297653/16095688400 bs1048576 nl0 frs1048576}
20:57:59.982649 tx 3:     OK, {blocks (54541,54286)/1571844 files 16092297653/16095688400 bs1048576 nl0 frs1048576}
20:57:59.982728 rx 2: STATFS n1 
20:57:59.982764 tx 2:     OK, {blocks (54541,54286)/1571844 files 16092297653/16095688400 bs1048576 nl0 frs1048576}
20:57:59.982817 rx 3: STATFS n1 
20:57:59.982837 tx 3:     OK, {blocks (54541,54286)/1571844 files 16092297653/16095688400 bs1048576 nl0 frs1048576}
20:57:59.983782 rx 2: LOOKUP n1 [".go-fuse-epoll-hack"] 20b
20:57:59.983841 tx 2:     OK, {n18446744073709551615 g0 tE=0s tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
20:57:59.983852 rx 3: INTERRUPT n0 {ix 2} 
20:57:59.983961 rx 4: GETATTR n18446744073709551615  
20:57:59.983971 tx 3:     35=resource temporarily unavailable
20:57:59.983996 tx 4:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
20:57:59.984080 rx 4: GETATTR n18446744073709551615  
20:57:59.984110 tx 4:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
20:57:59.984177 rx 4: GETATTR n18446744073709551615  
20:57:59.984214 tx 4:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
20:57:59.984284 rx 4: GETATTR n18446744073709551615  
20:57:59.984323 tx 4:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
20:57:59.984376 rx 4: GETATTR n18446744073709551615  
20:57:59.984404 tx 4:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
20:57:59.984462 rx 4: OPEN n18446744073709551615 {RDWR} 
20:57:59.984479 tx 4:     OK, {Fh 18446744073709551615 }
20:57:59.984559 rx 4: SETATTR n18446744073709551615 {size 0, fh 18446744073709551615} 
20:57:59.984596 tx 4:     34=result too large, {tA=0s {M00 SZ=0 L=0 0:0 0 0:0 A 0.000000 M 0.000000 C 0.000000}}
20:57:59.984645 rx 4: FLUSH n18446744073709551615 {Fh 18446744073709551615} 
THIREUS: IS THIS EVEN EXECUTED?
20:57:59.984672 tx 4:     OK
20:57:59.984729 rx 4: RELEASE n18446744073709551615 {Fh 18446744073709551615 RDWR  L0} 
THIREUS: IS THIS EVEN EXECUTED?
20:57:59.984754 tx 4:     OK
fs.Mount failed: result too large
```

**Comment by Thireus** [comment]
I've also tried to force the SETATTR status to be OK:
```
	case _OP_ACCESS, _OP_FLUSH, _OP_RELEASE, _OP_SETATTR:
		// Avoid upsetting the OSX mount process.
		req.status = OK
```

But it won't go further:
```
23:09:43.705710 rx 2: INIT n0 {7.19 Ra 0x200000 CASE_INSENSITIVE,ATOMIC_O_TRUNC,VOL_RENAME,XTIMES,0x18000000} 
23:09:43.705809 tx 2:     OK, {7.8 Ra 0x200000  0/0 Wr 0x20000 Tg 0x0}
23:09:43.705938 rx 2: STATFS n1 
23:09:43.705949 rx 3: STATFS n1 
23:09:43.706034 tx 2:     OK, {blocks (54621,54366)/1571844 files 16092300147/16095688400 bs1048576 nl0 frs1048576}
23:09:43.706059 tx 3:     OK, {blocks (54621,54366)/1571844 files 16092300147/16095688400 bs1048576 nl0 frs1048576}
23:09:43.706261 rx 2: STATFS n1 
23:09:43.706297 rx 3: STATFS n1 
23:09:43.706301 tx 2:     OK, {blocks (54621,54366)/1571844 files 16092300147/16095688400 bs1048576 nl0 frs1048576}
23:09:43.706337 tx 3:     OK, {blocks (54621,54366)/1571844 files 16092300147/16095688400 bs1048576 nl0 frs1048576}
23:09:43.706421 rx 2: STATFS n1 
23:09:43.706440 tx 2:     OK, {blocks (54621,54366)/1571844 files 16092300147/16095688400 bs1048576 nl0 frs1048576}
23:09:43.706484 rx 3: STATFS n1 
23:09:43.706518 tx 3:     OK, {blocks (54621,54366)/1571844 files 16092300147/16095688400 bs1048576 nl0 frs1048576}
23:09:43.707318 rx 2: LOOKUP n1 [".go-fuse-epoll-hack"] 20b
23:09:43.707367 tx 2:     OK, {n18446744073709551615 g0 tE=0s tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
23:09:43.707450 rx 3: GETATTR n18446744073709551615  
23:09:43.707513 tx 3:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
23:09:43.707636 rx 2: GETATTR n18446744073709551615  
23:09:43.707697 tx 2:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
23:09:43.707749 rx 3: GETATTR n18446744073709551615  
23:09:43.707769 tx 3:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
23:09:43.707827 rx 2: ACCESS n18446744073709551615 {u=0 g=0 w,r} 
23:09:43.707836 tx 2:     OK
23:09:43.707874 rx 3: GETATTR n18446744073709551615  
23:09:43.707891 tx 3:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
23:09:43.707935 rx 2: OPEN n18446744073709551615 {RDWR} 
23:09:43.707948 tx 2:     OK, {Fh 18446744073709551615 }
23:09:43.708018 rx 3: SETATTR n18446744073709551615 {size 0, fh 18446744073709551615} 
23:09:43.708053 tx 3:     OK, {tA=0s {M00 SZ=0 L=0 0:0 0 0:0 A 0.000000 M 0.000000 C 0.000000}}
23:09:43.708102 rx 2: FLUSH n18446744073709551615 {Fh 18446744073709551615} 
23:09:43.708122 tx 2:     OK
23:09:43.708160 rx 3: RELEASE n18446744073709551615 {Fh 18446744073709551615 RDWR  L0} 
23:09:43.708170 tx 3:     OK
fs.Mount failed: resource temporarily unavailable
```

This time I'm not sure why I see this error.

**Comment by hanwen** [comment]
I don't know either. Could you try commenting out the pollhack altogether? It looks like it might be something completely different.

**Comment by Thireus** [comment]
@hanwen, getting rid of pollhack for macOS should work just fine.

For now I'm using this hack, which you've said renders pollhack completely useless because the ".go-fuse-epoll-hack" file is created outside the mount:
`fd, err := syscall.Open(filepath.Join(filepath.Join(mountPoint, "/.."), pollHackName+"."+filepath.Base(mountPoint)), syscall.O_CREAT|syscall.O_TRUNC|syscall.O_RDWR, 0644)`

go-fuse works well with the above.

So, I would suggest disabling pollhack completely for macOS, until a pollhack compatibility with the latest version of OSXFuse is implemented.

**Comment by hanwen** [comment]
> 23:09:43.708018 rx 3: SETATTR n18446744073709551615 {size 0, fh 18446744073709551615} 
> 23:09:43.708053 tx 3:     OK, {tA=0s {M00 SZ=0 L=0 0:0 0 0:0 A 0.000000 M 0.000000 C 0.000000}}

hm, this looks invalid. The second 0:0 should be  0:18446744073709551615

**Comment by hanwen** [comment]
> go-fuse works well with the above.

well, kind of. You'll get random lockups if you try to unittest your file system.

**Comment by hanwen** [comment]
does this work https://review.gerrithub.io/c/hanwen/go-fuse/+/507313 ?

**Comment by Thireus** [comment]
@hanwen, that patch didn't work either. Same issue.

```
...
20:57:41.903746 rx 6: ACCESS n18446744073709551615 {u=502 g=20 w,r} 
20:57:41.903801 tx 6:     34=result too large
20:57:41.903817 rx 4: ACCESS n1 {u=502 g=20 } 
20:57:41.903936 rx 7: GETATTR n18446744073709551615  
20:57:41.903940 tx 4:     OK
20:57:41.903973 tx 7:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
20:57:41.904083 rx 2: ACCESS n1 {u=502 g=20 } 
fs.Mount failed: result too large
Maybe you should run: /Library/Filesystems/osxfuse.fs/Contents/Resources/load_osxfuse
20:57:41.904158 tx 2:     OK
```

> well, kind of. You'll get random lockups if you try to unittest your file system.

I see :(

**Comment by Thireus** [comment]
> does this work https://review.gerrithub.io/c/hanwen/go-fuse/+/507313 ?

Hey @hanwen, that seems to work! (I've deleted my previous reply, because I had not applied the correct patch)

```
21:10:58.095417 rx 2: INIT n0 {7.19 Ra 0x200000 XTIMES,VOL_RENAME,CASE_INSENSITIVE,ATOMIC_O_TRUNC,0x18000000} 
21:10:58.095520 tx 2:     OK, {7.8 Ra 0x200000  0/0 Wr 0x20000 Tg 0x0}
21:10:58.095599 rx 3: STATFS n1 
21:10:58.095621 rx 2: STATFS n1 
21:10:58.095667 tx 3:     OK, {blocks (7566369,2712984)/7633333 files 78163649210/78165332360 bs1048576 nl0 frs1048576}
21:10:58.095688 tx 2:     OK, {blocks (7566369,2712984)/7633333 files 78163649210/78165332360 bs1048576 nl0 frs1048576}
21:10:58.095912 rx 2: STATFS n1 
21:10:58.095917 rx 3: STATFS n1 
21:10:58.095957 tx 2:     OK, {blocks (7566369,2712984)/7633333 files 78163649210/78165332360 bs1048576 nl0 frs1048576}
21:10:58.095988 tx 3:     OK, {blocks (7566369,2712984)/7633333 files 78163649210/78165332360 bs1048576 nl0 frs1048576}
21:10:58.096039 rx 2: STATFS n1 
21:10:58.096069 tx 2:     OK, {blocks (7566369,2712984)/7633333 files 78163649210/78165332360 bs1048576 nl0 frs1048576}
21:10:58.096159 rx 3: STATFS n1 
21:10:58.096197 tx 3:     OK, {blocks (7566369,2712984)/7633333 files 78163649210/78165332360 bs1048576 nl0 frs1048576}
21:10:58.097013 rx 2: ACCESS n1 {u=502 g=20 x} 
21:10:58.097126 tx 2:     OK
21:10:58.097224 rx 3: LOOKUP n1 [".go-fuse-epoll-hack"] 20b
21:10:58.097293 tx 3:     OK, {n18446744073709551615 g0 tE=0s tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
21:10:58.097373 rx 2: GETATTR n18446744073709551615  
21:10:58.097415 tx 2:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
21:10:58.097478 rx 3: GETATTR n18446744073709551615  
21:10:58.097523 tx 3:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
21:10:58.097594 rx 2: GETATTR n18446744073709551615  
21:10:58.097633 tx 2:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
21:10:58.097724 rx 3: ACCESS n18446744073709551615 {u=502 g=20 w,r} 
21:10:58.097746 tx 3:     OK
21:10:58.097824 rx 2: GETATTR n18446744073709551615  
21:10:58.097866 tx 2:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
21:10:58.097927 rx 3: OPEN n18446744073709551615 {RDWR} 
21:10:58.097942 tx 3:     OK, {Fh 18446744073709551615 }
21:10:58.098022 rx 2: SETATTR n18446744073709551615 {size 0, fh 18446744073709551615} 
21:10:58.098052 tx 2:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
21:10:58.098101 rx 3: GETATTR n18446744073709551615  
21:10:58.098129 tx 3:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
21:10:58.098201 rx 2: FLUSH n18446744073709551615 {Fh 18446744073709551615} 
21:10:58.098219 tx 2:     OK
21:10:58.098263 rx 3: RELEASE n18446744073709551615 {Fh 18446744073709551615 RDWR  L0} 
21:10:58.098281 tx 3:     OK
Filesystem mounted and ready.
21:10:58.099052 rx 2: ACCESS n1 {u=0 g=0 } 
21:10:58.099140 tx 2:     OK
21:10:58.099203 rx 3: GETATTR n1  
21:10:58.099281 tx 3:     OK, {tA=1s {M040755 SZ=128 L=4 502:0 0 0:0 A 1607807451.125087 M 1607807453.476271 C 1607807453.476271}}
21:10:58.099390 rx 2: ACCESS n1 {u=502 g=20 } 
21:10:58.099390 rx 3: STATFS n1 
21:10:58.099462 tx 2:     OK
21:10:58.099484 tx 3:     OK, {blocks (7566369,2712984)/7633333 files 78163649210/78165332360 bs1048576 nl0 frs1048576}
21:10:58.099535 rx 2: ACCESS n1 {u=502 g=20 } 
21:10:58.099598 tx 2:     OK
21:10:58.099685 rx 3: ACCESS n1 {u=502 g=20 } 
21:10:58.099691 rx 2: STATFS n1 
21:10:58.099733 tx 3:     OK
21:10:58.099744 tx 2:     OK, {blocks (7566369,2712984)/7633333 files 78163649210/78165332360 bs1048576 nl0 frs1048576}
21:10:58.099804 rx 3: ACCESS n1 {u=502 g=20 } 
```

**Comment by Thireus** [comment]
Although... the examples seem to be broken:

```
bash-3.2$ go run main.go /tmp/z foo.zip 
go: downloading golang.org/x/sys v0.0.0-20180830151530-49385e6e1522
# github.com/hanwen/go-fuse/v2/zipfs
../../zipfs/zipfs.go:88:5: out.Blksize undefined (type *fuse.AttrOut has no field or method Blksize)
bash-3.2$ cd ../hello/
bash-3.2$ go run main.go tmp
2020/12/12 21:14:38 Mount fail: result too large
exit status 1
```

But I can confirm with your latest patch I can mount and use gocryptfs fuse volumes just fine.

**Comment by john-dev** [comment]
I can confirm the patch does fix it indeed for me too on MacOS 10.15.6

**Comment by hanwen** [comment]
submitted as https://github.com/hanwen/go-fuse/commit/9473722efcb1906b8afbecb1a7ba9d33b34f09a7
